### ISSUE TITLE
Bias in EvalCollector

### ISSUE THREAD

**Issue by kaixin96** [astooke/rlpyt#93]
`EvalCollector` (e.g. `GpuEvalCollector`) seems to bias towards short trajectories, leading to inaccurate evaluation. The problem occurs when `eval_n_envs > 1`.

In [GpuEvalCollector.collect_evaluation()](https://github.com/astooke/rlpyt/blob/master/rlpyt/samplers/parallel/gpu/collectors.py#L125), the list of `TrajInfoCls` is created as a local variable. Unfinished trajectories in `traj_infos` will be discarded at the end of this evaluation. Next evaluation will start with another empty `traj_infos`. In consequence,  short trajectories (usually of low reward) are more likely to be collected.

In my experiment, I tested the trained model offline with only 1 env and got almost twice the reward in the eval logs. This can also be confirmed in the average_eval_steps statistic. In my offline testing the average_eval_steps is ~600 while in the logs average_eval_steps is only ~100.

Although the severity of this bias is also influenced by some hyper-parameters (e.g. `eval_n_envs`, `eval_max_steps`), it is a general bias unless `eval_n_envs=1`.

A simple fix is to add `traj_infos` as an attribute of `EvalCollector` (that is, using `self.traj_infos` instead). I can send a PR.

Thank you.

**Comment by astooke** [comment]
Hi, this is a good thing to point out.  The approach I've usually taken is just to make the total amount of eval trajectories collected larger (not ideal if it takes a long time).  Or for example, we'll make sure that at least as many steps are taken (`eval_max_steps` / `eval_n_envs`) as the environment horizon (if there is one), so that every trajectory in the starting set definitely finishes.

It might also skew the results to use partially finished trajectories, which we usually wanted to avoid.  I guess in your case this is better to do?

Watch out, the number of `eval_n_envs` might get rounded up to one per worker?

**Comment by kaixin96** [comment]
Hi @astooke, thanks for your reply. I am afraid the evaluation result is still skewed even though many steps are taken. Making sure that every trajectory in the starting set definitely finishes can mitigate the issue but can't solve it. We still tend to have more short trajectories. In my case, I also want to avoid skewed evaluation results. `eval_n_envs` is larger than 1.

Some solutions I can think of:
- maintain `eval_traj_infos` as done for `train_traj_infos` and do not reset at the start of `collect_evaluation()`
- make sure that equal episodes are sampled in each eval env (this might affect parallel efficiency)

Thank you.

**Comment by astooke** [comment]
Yes I think the one that is most correct is to designate ahead of time that each eval env will run exactly a fixed number of episodes, even if that diminishes the parallel efficiency when some eval workers have short trajectories and finish early.  Hmm that sounds like a really useful option to put in, as we have sometimes worried about this, too.

**Comment by kaixin96** [comment]
I agree that running same fixed number of episodes for each eval env is most correct.
